decoder.go

```diff
@@ -54,6 +54,15 @@ func (r *Decoder) readByte() (byte, error) {
 	return 0, err
 }
 
+// readIn takes a []byte to fill it and check errors appropriately
+func (r *Decoder) readIn(data []byte) error {
+	n, err := r.r.Read(data)
+	if n == len(data) {
+		return nil
+	}
+	return err
+}
+
 func (r *Decoder) readSlice(delim byte) (data []byte, err error) {
```

Comment:
this can be renamed `readByteSliceUntil` since now there is a proper `readByteSlice`